group/assignment-6/src/cars/RearrangingCars.java

```diff
@@ -0,0 +1,323 @@
+package cars;
+
+import cars.parkinglot.CarMove;
+import cars.parkinglot.ParkingLot;
+import cars.parkinglot.ParkingLotConfigurations;
+import cars.parkinglot.ParkingSpot;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+public final class RearrangingCars {
+
+    /**
+     * Static utility class, no need to create an instance.
+     */
+    private RearrangingCars() {
+    }
+
+    public static List<CarMove> findSequenceOfMoves(ParkingLot startState, ParkingLot endState)
+            throws IllegalArgumentException {
+        validateParkingLots(startState, endState);
+
+        List<CarMove> carMoves = new ArrayList<>();
+        Set<Car> carsToBeMoved = findCarsToBeMoved(startState, endState);
+        ParkingSpot defaultSpotFrom = endState.getEmptySpot();
+        ParkingSpot currentEmptySpot = startState.getEmptySpot();
+        @Nullable
+        Car nextCarToMove = endState.getCarAt(currentEmptySpot);
+        ParkingSpot spotFrom;
+
+        while (!carsToBeMoved.isEmpty()) {
+            // nextCarToMove == null when the current empty spot is the same one as that in the end state
+            while (nextCarToMove != null) {
+                // every car moves up to 2 times.
+                // if a car's is moved for the second time it won't be contained in carsToBeMoved,
+                // and it would mean that its first move was a temporary one to the end state's empty spot
+                // (defaultSpotFrom)
+                spotFrom =
+                        carsToBeMoved.contains(nextCarToMove)
+                                ? startState.getSpotOf(nextCarToMove)
+                                : defaultSpotFrom;
+                carsToBeMoved.remove(nextCarToMove);
+                carMoves.add(new CarMove(nextCarToMove, spotFrom, currentEmptySpot));
+                currentEmptySpot = spotFrom;
+                nextCarToMove = endState.getCarAt(currentEmptySpot);
+            }
+
+            if (!carsToBeMoved.isEmpty()) {
+                nextCarToMove = carsToBeMoved.iterator().next();
+            }
+        }
+
+        return carMoves;
+    }
+
+    public static List<List<CarMove>> getAllSequencesOfMoves(ParkingLot startState, ParkingLot endState)
+            throws IllegalArgumentException {
+        validateParkingLots(startState, endState);
+        Map<ParkingSpot, Car> configuration = startState.getParkingLotState();
+        ParkingLotConfigurations<ParkingSpot> treeOfConfigurations = new ParkingLotConfigurations(configuration.values());
```

Comment:
```suggestion
        ParkingLotConfigurations<ParkingSpot> treeOfConfigurations = new ParkingLotConfigurations<>(configuration.values());
```

same as in `findSequenceOfMoves`